nfs/src/rpc_task.cpp

```diff
@@ -476,6 +476,377 @@ void rpc_task::run_setattr()
     } while (rpc_retry);
 }
 
+void rpc_task::run_read()
+{
+    auto ino = rpc_api.read_task.get_inode();
+
+    auto readfile_handle =
+        get_client()->get_nfs_inode_from_ino(ino)->filecache_handle;
+
+    bytes_vector = readfile_handle->get(
+                       rpc_api.read_task.get_offset(),
+                       rpc_api.read_task.get_size());
+
+    /*
+     * Now go through the byte chunk vector to see if the buffers are
+     * populated. If the chunks are empty, we will issue parallel read
+     * calls to fetch the data.
+     */
+    const size_t size = bytes_vector.size();
+
+    // There should not be any reads running for this RPC task initially.
+    assert (num_of_reads_issued_to_backend == 0);
+
+    AZLogDebug("run_read:: offset {}, size: {}, num_of_chunks: {}",
+               rpc_api.read_task.get_offset(),
+               rpc_api.read_task.get_size(),
+               size);
+
+    /*
+     * Check if the requested data is present in the cache.
+     * If yes, send response to caller by reading data from cache.
+     */
+    size_t data_length_in_cache = 0;
+    for (size_t i = 0; i < size; i++) {
+        // If the data is present in cache, membuf should be marked uptodate.
+        if (bytes_vector[i].get_membuf()->is_uptodate())
+        {
+            // Data exists in the cache, we need not issue backend read.
+            data_length_in_cache += bytes_vector[i].length;
+            if (data_length_in_cache >= rpc_api.read_task.get_size())
```

Comment:
since get() will return enough bytes_chunk to cover the requested range, we should never have a case where this condition is hit before we exhaust all the bytes_chunk returned.